config/config_koanf_hooks.go

```diff
@@ -0,0 +1,171 @@
+package config
+
+import (
+	"encoding"
+	"fmt"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/miekg/dns"
+	"github.com/mitchellh/mapstructure"
+)
+
+// nolint:gochecknoglobals
+var (
+	koanfHookTypes = []reflect.Type{
+		reflect.TypeOf(Duration(0)),
+		reflect.TypeOf(Upstream{}),
+		reflect.TypeOf(QTypeSet{}),
+	}
+)
+
+const (
+	queryTypeMax = 110
+)
+
+func hasCustomHook(t reflect.Type) bool {
+	for _, kht := range koanfHookTypes {
+		if kht == t {
+			return true
+		}
+	}
+
+	return false
+}
+
+func queryTypeHookFunc() mapstructure.DecodeHookFuncType {
+	return func(
+		f reflect.Type,
+		t reflect.Type,
+		data interface{}) (interface{}, error) {
+		if f.Kind() == reflect.Slice &&
+			t == reflect.TypeOf(QTypeSet{}) {
+			s := reflect.ValueOf(data)
```

Comment:
I think you can do this to get an actual slice: `s := reflect.ValueOf(data).Interface().([]any)`  
That would make the iteration work with the classic `range` which is nicer IMO.